<a href="https://colab.research.google.com/github/rajkstats/uplimit-llm-in-production/blob/main/LLMS_In_Production_W1_Project_FINAL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DataAlchemy Labs


<img src="https://drive.google.com/uc?id=1jLQB9IRtp4cHnIxRSebSVcclKxNfdHCf" />

Welcome to the week 1 project for LLMs in Production. In this weeks project you are a founding Machine Learning Engineer on the team you and your team have narrowed down a life-changing product. You will be building and launching your groundbreaking SaaS product which is quite similar to a product already out in the market, [AI2sql](https://www.ai2sql.io/).

![AI2sql Example](https://drive.google.com/uc?id=1mWOQs6OZmuCLui-auLosXsXGhwl2mbVg)

You have done a significant amount of market research and are confident that this product will gather a cult-like following with a killer feature of being able to take any natural language query and output SQL code based on the query.

Potential users of your application are:
- **Project Managers**: They can use our tool to quickly derive important metrics for presentations or reports. Instead of relying on technical teams, they can directly query the database using natural language.
- **Data Scientists**: Often dealing with complex data queries, data scientists can use our tool to streamline and debug intricate SQL statements, saving time and reducing the potential for error.
- **Business Analysts and Non-Tech Professionals**: Anyone without deep SQL knowledge but needing to interact with the database can benefit immensely. They can easily convert their data needs into SQL queries without delving into the complexities of SQL syntax.

There are many more features that you can work on in the future such as:
1. NoSQL Code Generation
2. SQL Syntax Checking
3. Explaining SQL Code
4. Optimizing SQL Code
5. Formatting SQL Code

But since this is an MVP and you want to laucnh as quickly as possible we will be just focussing on the use case of SQL Code Generation. In particular we will be supporting the use of **SQL** initially as its the most popular database out there in use currently based on the recent [StackOverflow Developer Survey 2023](https://survey.stackoverflow.co/2023/#section-most-popular-technologies-databases).



Our final product will be a simple web application with a basic Q/A system that allows users to ask query about SQL and then get back SQL code that they can use for their application, think of it as a super basic version of the ChatGPT from OpenAI that I am sure everyone is quite familiar with.

![ChatGPT OpenAI Example](https://drive.google.com/uc?id=1AGVZoxtWvBF6KLX0mkvpnU7mMEgaSraa)

In the Week 1 Project we will be solidfying what we learnt throughout the week by building out this core functionality for the SQL Code Generation part of our application. Just like all Data Science projects we start off in a notebook environment to prototype things before having to bring them to production.

Just like most LLM startups we will be leveraging OpenAI and their ChatGPT API initially for our product, in particular we will be using **gpt-4o-mini** which offers a good enough LLM that is quite fast!

The main parts that we will be covering in the project are:
1. [Using LLMs to Generate SQL Code](#scrollTo=HcGIbmPiM4o3)
2. [Evaluating LLMs on SQL Code Generation](#scrollTo=pWz1btwOM7zr)
3. [Validating LLM Outputs using Guardrails](#scrollTo=rPmPTQssNDUn)
4. [Extra Credit: Optional Tasks](#scrollTo=UL-XI8PiNU1z&line=1&uniqifier=1)

<a href="https://colab.research.google.com/drive/10y3EFKr4S8TsiSp00y8_LTD-2UU8nCGS?usp=sharing" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>





# Setting up OpenAI API Key

<div style="
  padding: 10px;
  border-radius: 5px;
  background-color: #ffcccc;
  border-left: 6px solid #ff0000;
  margin-bottom: 20px;">
  
  <strong>⚠️ Important Notice:</strong>
  <p>Do not share or use this API Key outside of the context of the notebook exercises.</p>
</div>

Uplimit has provisioned an OpenAI API Key for your projects. Please add this API Key to this assignment by clicking on the Security Key icon on the left hand tab of the Google Colab notebook and then add a new parameter value called `OPENAI_API_KEY`.


 Here you can provide the API key that you copied and this will not be part of your Google Colab account. You can also enable the toggle Notebook access - this will allow your notebook to have access to this API key.

<img src="https://drive.google.com/uc?id=1SfE-nNOQ3DfZpJN6mAnxuqVsMv-F1_Yp" />

**NOTE:** We are hardcoding an API key for [Guardrails.ai](https://www.guardrailsai.com/) which is used for downloading different types of LLM Guardrails, access is free regardless and we hardcode it as a convenience for all the students.

After the API Key has been setup, run the following code:

In [3]:
!pip install openai guardrails-ai==0.5.0 deepeval langchain langchain-openai

In [4]:
from openai import OpenAI
from google.colab import userdata

import os
os.environ["OPENAI_API_KEY"] = userdata.get('UL_OPENAI_API_KEY')
os.environ["GUARDRAILS_TOKEN"] = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJnb29nbGUtb2F1dGgyfDEwNzI3MDE2NDYwNDM4MTg2OTU4OSIsImFwaUtleUlkIjoiYjAzNDVkNGEtZDhjNy00OThmLWIwZGYtZWI3ZTY0MzMwNTJkIiwiaWF0IjoxNzMwMTQ2ODc0LCJleHAiOjE3Mzc5MjI4NzR9.1J84O3pT_KlWTkGjV4zfBMChqYow868A-XTsePwiZ_Q"

# 1. Using LLMs to Generate SQL Code

## SQL Code Generation

SQL is one of the most powerful programming languages out there and is the main way in which we communicate with databases. Without it backend engineers would be unable to store and structure data to be easily queried to return to a frontend application and data scientists would be unable to track key metrics about their experiments over time.

SQL is the language used for interacting with databases, while SQL is a specific database system that understands and uses SQL for database operations. Each database system, including SQL, implements SQL with some variations and adds its own proprietary extensions to the standard SQL language.

Here is an example SQL statement:

```sql
SELECT
    u.name,
    u.email,
    COUNT(o.order_id) AS total_orders,
    SUM(o.amount) AS total_amount_spent
FROM
    users u
JOIN
    orders o ON u.user_id = o.user_id
WHERE
    o.order_date >= (NOW() - INTERVAL '1 year')
GROUP BY
    u.user_id
HAVING
    COUNT(o.order_id) > 5
ORDER BY
    total_amount_spent DESC;
```

LLM's such as ChatGPT from OpenAI are trained on all kinds of data across the internet and in particular they are trained on code across many different programming languages.


![Text2SQL](https://drive.google.com/uc?id=17IDU11L_JdPIfMWvDa6li_5oOyGgWriX)

What we will be performing is asking the LLM a Question related to a SQL query in order to create SQL code that we can then run.


During OpenAI DevDay, there was a fantastic breakout section about **"[A Survey of Techniques for Maximizing LLM Performance](https://youtu.be/ahnGLM-RC1Y?si=rifoSUxFgvgliFtT)"** which we recommend all of you go through. In that talk, the speakers highlighted how we should always be starting off with Prompt Engineering when building LLM applications. Once we understand what our LLM application lacks can we only try more complex techniques such as RAG or Fine-tuning or both. With that spirit in mind let's try out some basic Prompt Engineering using ChatGPT.

![Optimization Flow from OpenAI](https://drive.google.com/uc?id=1kcguLT8KYBmDypGB0fVyeT8YfrCiiMx_)

In [5]:
from langchain_openai import ChatOpenAI
from langchain.schema import LLMResult, HumanMessage, Generation

MODEL_NAME = "gpt-4o-mini"

llm = ChatOpenAI(
    model_name=MODEL_NAME,
    temperature=0.0
)

## TODO: Create a Baseline Zero-Shot Prompt

Create a Zero-Shot Prompt that we can use in this situation, we have provided you with a simple prompt to try out. A **Zero-Shot prompt** is where we are asking the model to perform a task it has never seen explicty during training without any additional examples.

When working on the prompt or improving the prompt remember a couple of things:

1. Make sure to use clear instructions
2. Try splitting tasks into simpler subtasks
3. Give the model time to “think”

![How to Prompt Engineer](https://drive.google.com/uc?id=1iSZ6XvD5eMiRSQBUT0DFEDA-1qc0qibv)


In [7]:
# Add your prompt here, make sure to use
# {query} to inject a query into the prompt.

ZERO_SHOT_PROMPT_TEMPLATE = """
You are an expert SQL developer tasked with translating natural language queries into precise SQL statements. Follow these steps:

Analyze the query:
- Identify the main tables involved
- Determine the required operations (SELECT, JOIN, GROUP BY,ORDER BY etc.)
- Note any conditions or constraints

ASSUMPTIONS:
- List any assumptions about table structures and relationships
- If column names aren't specified, use common conventional names
- Document any assumed data types

SQL Translation:

Write the SQL query following these guidelines:
- Use proper SQL syntax and formatting
- Include appropriate table aliases where needed
- Add comments to explain complex logic
- Consider performance implications

Explanation:
- Provide a brief explanation of how the query works

Here is the query to translate: {query}

Please format your response as follows:

SQL QUERY:
[Your SQL query here]

Assumtions:
[List your assumptions here]

Explanation:
[Brief explanation of the query logic]
"""

# Try your own query
SAMPLE_QUERY = "Select the name of the employee who has the highest salary"

prompt = ZERO_SHOT_PROMPT_TEMPLATE.format(
    query=SAMPLE_QUERY
)

result = llm.generate([[HumanMessage(content=prompt)]])

# # Do try to investigate the generations from LangChain to understand what is generated
print(result.generations[0][0].text)

SQL QUERY:
```sql
SELECT e.name
FROM employees e
WHERE e.salary = (SELECT MAX(salary) FROM employees)
```

Assumptions:
1. There is a table named `employees` that contains employee data.
2. The `employees` table has at least the following columns:
   - `name` (data type: VARCHAR or similar)
   - `salary` (data type: DECIMAL or FLOAT)
3. The `salary` column contains numeric values representing the employee's salary.
4. It is assumed that there is at least one employee in the `employees` table.

Explanation:
The query retrieves the name of the employee with the highest salary. It does this by using a subquery to find the maximum salary from the `employees` table. The outer query then selects the name of the employee whose salary matches this maximum value. This approach ensures that if there are multiple employees with the same highest salary, all of their names will be returned.


Great job! We have generated our very first SQL code using an LLM!

An improvement we can perform on this is using a **Few-Shot prompt** where we include a couple of examples for the LLM to understand what we are trying to do. Doing so should result in better results as the LLM has some reference examples about what we are trying to do.

**BUT** remember *we need to evaluate our LLM* prompt before we can even think about experimenting on it otherwise how would we even know if this new prompt is even better in the first place?

# 2. Evaluating LLMs on SQL Code Generation

Evaluating Text-to-SQL might seem like an easy tasks to evaluate as all we need in order to generate our Golden Dataset would be a list of natural language queries and their respective SQL query but there are some intricacies with it that will prove to be challenging.

We will need some way to check for **equivalence** to the **gold** query, which is where prior mentioned intricacies will start to pop up, but we will circle back to them later on. With that, we can start off with looking for an exact match between the LLM response and the golden query. For our initial metric we will be using a very simple metric, [exact match](https://huggingface.co/spaces/evaluate-metric/exact_match).

![Basic Exact Match LLM Evaluator](https://drive.google.com/uc?id=1yIygL2zNC4JZ02dMBtC5HyNa0_ujIOOv)

With a Golden Dataset which we can hand generate using our domain expertise, we can build up an evaluation dataset to run on every iteration to improve our application. Using the exact match metric we can derive an accuracy score of our LLM-application that we can keep track of and use in order to compare the performance across different iterations overtime.

Here is an example of what our dataset can look like:

![Example Golden Dataset](https://drive.google.com/uc?id=12D8C2juErtOzvKVLaKL_njxlceFJcbTu)

**NOTE:** If you noticed, there is a **complexity** column on the SQL code used in the example which we can try incorporating as an optional tasks in order to improve the quality of our results across different fine-grained categories which is paramount in ensuring our LLM SQL Code Generator is more robust. Ideally we should have an even distribution across all these additional categories that we add on, where we can then calculate metrics such as **Accuracy@HardComplexity** and so on.

We will provide you with a basic evaluation dataset to start off with, which we can further improve on using the process laid out below using [deepeval](https://github.com/confident-ai/deepeval):

![Evaluation Process](https://drive.google.com/uc?id=1KC_gMvU3PeflEm0N1Co8_YnT46lCvMLU)
<!-- 1. Load a Golden Test set for the students in a JSONL file with 20-30 examples
2. TODO: Create a custom deep-eval evaluator for SQL code where we check if the output of the SQL code exactly matches the output from the LLM.
3. TODO: Run evaluation on the golden test set.
4. TODO: Try different prompts and see if they improve the evaluation metric
5. Checking for an exact match is not the best way to go about it as the same query could have multiple ways to do it. Let's try using G-Eval where we use another more powerful LLM.
6. TODO: Evaluate using G-Eval
7. Talk about how in a Production system we should be running evaluation just like how we run automated tests, via CI/CD pipelines. -->

In [8]:
BASIC_GROUND_TRUTH_DATASET = [
  {
    "Query": "Show the total number of employees.",
    "Ground Truth": "SELECT COUNT(*) AS total_employees FROM employees;"
  },
  {
    "Query": "Find the highest salary in the Finance department.",
    "Ground Truth": "SELECT MAX(salary) FROM employees WHERE department = 'Finance';"
  },
  {
    "Query": "Get the average age of all managers.",
    "Ground Truth": "SELECT AVG(age) FROM employees WHERE position = 'Manager';"
  },
  {
    "Query": "List the names and emails of staff in the IT department.",
    "Ground Truth": "SELECT name, email FROM employees WHERE department = 'IT';"
  },
  {
    "Query": "What are the titles of the top 5 selling books?",
    "Ground Truth": "SELECT title FROM books ORDER BY sales DESC LIMIT 5;"
  },
  {
    "Query": "Which product has the least quantity in stock?",
    "Ground Truth": "SELECT product_name FROM products ORDER BY quantity_in_stock ASC LIMIT 1;"
  },
  {
    "Query": "Display the second highest salary in the organization.",
    "Ground Truth": "SELECT MAX(salary) FROM employees WHERE salary < (SELECT MAX(salary) FROM employees);"
  },
  {
    "Query": "List employees who joined after 2015 and work in the Sales department.",
    "Ground Truth": "SELECT * FROM employees WHERE year(joined_date) > 2015 AND department = 'Sales';"
  },
  {
    "Query": "Find the average order value for each customer.",
    "Ground Truth": "SELECT customer_id, AVG(order_value) FROM orders GROUP BY customer_id;"
  },
  {
    "Query": "Show departments that have more than 10 employees.",
    "Ground Truth": "SELECT department FROM employees GROUP BY department HAVING COUNT(*) > 10;"
  },
  {
    "Query": "List all products that have never been ordered.",
    "Ground Truth": "SELECT * FROM products WHERE product_id NOT IN (SELECT product_id FROM orders);"
  },
  {
    "Query": "Which customers have spent more than $1000 in total?",
    "Ground Truth": "SELECT customer_id FROM orders GROUP BY customer_id HAVING SUM(order_value) > 1000;"
  },
  {
    "Query": "Show the total number of orders placed each day last week.",
    "Ground Truth": "SELECT order_date, COUNT(*) FROM orders WHERE order_date > CURRENT_DATE - INTERVAL '7 days' GROUP BY order_date;"
  },
  {
    "Query": "List the names of employees who do not manage anyone.",
    "Ground Truth": "SELECT name FROM employees WHERE name NOT IN (SELECT manager_name FROM employees);"
  },
  {
    "Query": "Display the department that has the highest average employee salary.",
    "Ground Truth": "SELECT department FROM employees GROUP BY department ORDER BY AVG(salary) DESC LIMIT 1;"
  }
]

In [9]:
import pandas as pd

df = pd.DataFrame(BASIC_GROUND_TRUTH_DATASET)
df.head()

Query  \
0                Show the total number of employees.   
1  Find the highest salary in the Finance departm...   
2               Get the average age of all managers.   
3  List the names and emails of staff in the IT d...   
4    What are the titles of the top 5 selling books?   

                                        Ground Truth  
0  SELECT COUNT(*) AS total_employees FROM employ...  
1  SELECT MAX(salary) FROM employees WHERE depart...  
2  SELECT AVG(age) FROM employees WHERE position ...  
3  SELECT name, email FROM employees WHERE depart...  
4  SELECT title FROM books ORDER BY sales DESC LI...

In the next step, lets see how our LLM and Prompt are able to perform on the entire dataset by running it through and evaluating its performance. This is known as performing a LLM Task Evaluation which is different than LLM Model Evals. Task evaluation cares about the performance of the LLM on a specific task whereas Model evals care about the general performance of the LLM across a variety of tasks, benchmarks such as the [HellaSwag](https://rowanzellers.com/hellaswag/) dataset is an example of a Model Evals and Text2SQL accuracy like we are doing in this project is an example of task evaluation.

We will start off with traditional NLP metrics that compare the output with the ground truth can derive a score based on the exact/similarity of the two. These metrics are simple to use and serve as a good place to start off when evaluating LLMs.

Firstly, let's start off with the exact match metric which very simply just compares the output with the ground truth and checks whether its an exact match or not, deepeval implements these metrics within the [Scorer module](https://github.com/confident-ai/deepeval/blob/4b3ceed20993232331550798fe0a8f1bf2605594/deepeval/scorer/scorer.py#L7).

deepeval makes it easy to create our own metrics by inheriting the BaseMetric class, you can read more about creating custom metrics [here](https://docs.confident-ai.com/docs/metrics-custom).

In [10]:
from deepeval.metrics import BaseMetric
from deepeval.test_case import LLMTestCase
from deepeval.scorer import Scorer


class ExactMatchMetric(BaseMetric):
    def __init__(self, threshold: float = 0.0, async_mode: bool = True):
        self.threshold = threshold
        self.async_mode = async_mode

    def _base_measure(self, test_case: LLMTestCase):
        self.success = Scorer.exact_match_score(
            test_case.actual_output, test_case.expected_output
        )
        if self.success:
            self.score = 1
        else:
            self.score = 0
        return self.score

    def measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    async def a_measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    def is_successful(self):
        return bool(self.success)

    @property
    def __name__(self):
        return "ExactMatch"

In [11]:
# Let's create our first test case!

test_case_input = df.head(1)
test_case_query = test_case_input['Query'].squeeze()
test_case_expected_output = test_case_input['Ground Truth'].squeeze()

In [12]:
print(f"Here is our query: {test_case_query}")
print(f"Here is our ground truth: {test_case_expected_output}")

Here is our query: Show the total number of employees.
Here is our ground truth: SELECT COUNT(*) AS total_employees FROM employees;


In [13]:
ZERO_SHOT_PROMPT_TEMPLATE = """
Generate a valid SQL query for the following natural language instruction:

${query}

Only generate SQL code and nothing else.
"""

prompt = ZERO_SHOT_PROMPT_TEMPLATE.format(
    query=test_case_query
)

result = llm.generate([[HumanMessage(content=prompt)]])

test_case_actual_output = result.generations[0][0].text

In [14]:
print(f"Here is what our model generated: {test_case_actual_output}")

Here is what our model generated: ```sql
SELECT COUNT(*) AS total_employees FROM employees;
```


In [15]:
from deepeval.test_case import LLMTestCase

test_case = LLMTestCase(
  input=test_case_input,
  actual_output="SELECT COUNT(*) AS total_employees FROM employees;",
  expected_output=test_case_expected_output
)

In [16]:
exact_match_metric = ExactMatchMetric()
exact_match_metric.measure(test_case)

In [17]:
print(f"Test case: {exact_match_metric.is_successful()}")

Test case: True


**NOTE:** If your prompt still fails and produces a different SQL code such as *"SELECT COUNT(*) AS total_employees FROM employees;"* feel free to update the ground truth to get it to pass in the meantime. This can happen as code correctness is not a simple task due to the fact that different implementations can still solve the query which is something we invite you to address via the optional tasks for this week.


Great job! We are able to see that our LLM and Prompt are able to answer the test case correctly, but that is just a basic example that we have. Next let us run it on the entire dataset that we have created.

## TODO: Create a deepeval dataset and evalute our LLM and Prompt on it

There is a slight difference between what we have been learning about Golden Datasets in general and how Golden Datasets work within deepeval. When we talk about a Golden Dataset as per the [course materials here](https://uplimit.com/course/llms-in-production/v2/module/evaluating-llms#corise_clrnjsvpd001m2e6iogarwdpn) we mention that its similar to a held-out test set meaning that we have data to predict on alongside their ground truth labels in order to derive evaluation metrics.

In deepeval, we will need to create an Evaluation Dataset instead where an Evaluation Dataset is built from LLMTestCase's and/or Golden's. Remember that the `actual_output` comes from running our LLM against the `input` as we did above!
```python
from deepeval.test_case import LLMTestCase
from deepeval.dataset import EvaluationDataset

first_test_case = LLMTestCase(input="...", actual_output="...")
second_test_case = LLMTestCase(input="...", actual_output="...")

test_cases = [first_test_case, second_test_case]
dataset = EvaluationDataset(test_cases=test_cases)
```
The definition of Golden's within deepeval is documented [here](https://docs.confident-ai.com/docs/confident-ai-manage-datasets#what-is-a-golden):
```
A "Golden" is what makes up an evaluation dataset and is very similar to a test case in deepeval, but they:

    - do not require an actual_output, so whilst test cases are always ready for evaluation, a golden isn't.
    - only exists within an EvaluationDataset(), while test cases can be defined anywhere.
    - contains an extra additional_metadata field, which is a dictionary you can define on Confident. Allows you to do some extra preprocessing on your dataset (eg., generating a custom LLM actual_output based on some variables in additional_metadata) before evaluation.

We introduced the concept of goldens because it allows you to create evaluation datasets on Confident without needing pre-computed actual_outputs. This is especially helpful if you are looking to generate responses from your LLM application at evaluation time.
```
With our EvaluationDataset created we can then run our LLM through it in order to derive the final score. For example, the following does so without using Pytest although its recommended that you treat this process as a CI/CD process that runs alongside your Pytest suite for your actual application.

```python
from deepeval import evaluate
from deepeval.metrics import HallucinationMetric, AnswerRelevancyMetric
from deepeval.dataset import EvaluationDataset

dataset = EvaluationDataset(test_cases=[...])
hallucination_metric = HallucinationMetric(threshold=0.3, context=["..."])
answer_relevancy_metric = AnswerRelevancyMetric(threshold=0.5)

dataset.evaluate([hallucination_metric, answer_relevancy_metric])

# You can also call the evaluate() function directly
evaluate(dataset, [hallucination_metric, answer_relevancy_metric])
```
An example output of the evaluation would look like this:
```
======================================================================

Metrics Summary

  - ✅ Answer Relevancy (score: 1, threshold: 0.5, strict: False, evaluation model: gpt-4-turbo, reason: The score is 1.00 because the response perfectly addresses the question about operating hours without any irrelevant information. Great job!, error: None)
  - ✅ Bias (score: 0, threshold: 0.5, strict: False, evaluation model: gpt-4-turbo, reason: The score is 0.00 because the output maintains neutrality and objectivity, as evidenced by the absence of any cited biased phrases or issues., error: None)

For test case:

  - input: What are your operating hours?
  - actual output: ...
  - expected output: None
  - context: ['Our company operates from 10 AM to 6 PM, Monday to Friday.', 'We are closed on weekends and public holidays.', 'Our customer service is available 24/7.']
  - retrieval context: None

======================================================================

Metrics Summary

  - ✅ Answer Relevancy (score: 1, threshold: 0.5, strict: False, evaluation model: gpt-4-turbo, reason: The score is 1.00 because the response accurately addresses the question about free shipping without any irrelevant information. Great job on maintaining focus!, error: None)
  - ✅ Bias (score: 0, threshold: 0.5, strict: False, evaluation model: gpt-4-turbo, reason: The score is 0.00 because the actual output perfectly demonstrates neutrality and objectivity, without any indication of bias., error: None)

For test case:

  - input: Do you offer free shipping?
  - actual output: ...
  - expected output: Yes, we offer free shipping on orders over $50.
  - context: None
  - retrieval context: None

======================================================================
```
One thing to note is that the metrics are run on a single (query, ground_truth) instead across the entire dataset meaning that we lack the ability to have a singular number to compare across different evaluation runs like traditional machine learning where you would for example have a accuracy score for the test set.

Let us introduce the concept of pass rates which are similar to accuracy across the entire dataset but its just meant to show how many instances for a metric in the dataset passed a evaluation metric check.

With this you could enforce a strict pass rate on deployments much like how we can enforce Pytest code coverage to be off a certain %.

deepeval supports this via the `aggregate_metric_pass_rates()` on the results of the evaluation give us pass rates across each metric that we use:

```python
from deepeval.evaluate import aggregate_metric_pass_rates

evaluation_results = evaluate(dataset, [hallucination_metric, answer_relevancy_metric])
aggregate_metric_pass_rates(evaluation_results.test_results)
```

This should output the following where we get the pass rates across each metric and a mapping storing the pass rates across each metric which we can use to store and compare across different evaluation runs:
```
======================================================================

Aggregate Metric Pass Rates

AnswerRelevancyMetric: 100.00% pass rate
BiasMetric: 100.00% pass rate

======================================================================

{'AnswerRelevancyMetric': 1.0, 'BiasMetric': 1.0}
```

Now with this you can apply it to our dataset:
1. Create an [EvaluationDataset](https://docs.confident-ai.com/docs/evaluation-datasets#create-an-evaluation-dataset) using deepeval.
2. Run the evaluation on the Golden Dataset without Pytest, refer to [this](https://docs.confident-ai.com/docs/evaluation-introduction#evaluating-without-pytest).

In [25]:
df.head()

Query  \
0                Show the total number of employees.   
1  Find the highest salary in the Finance departm...   
2               Get the average age of all managers.   
3  List the names and emails of staff in the IT d...   
4    What are the titles of the top 5 selling books?   

                                        Ground Truth  
0  SELECT COUNT(*) AS total_employees FROM employ...  
1  SELECT MAX(salary) FROM employees WHERE depart...  
2  SELECT AVG(age) FROM employees WHERE position ...  
3  SELECT name, email FROM employees WHERE depart...  
4  SELECT title FROM books ORDER BY sales DESC LI...

In [29]:
from deepeval import evaluate
from deepeval.test_case import LLMTestCase
from deepeval.dataset import EvaluationDataset
from deepeval.metrics import (
    HallucinationMetric,
    AnswerRelevancyMetric)
from deepeval.evaluate import aggregate_metric_pass_rates


class ExactMatchMetric(BaseMetric):
    def __init__(self, threshold=0.0, async_mode=True):
        self.threshold = threshold
        self.async_mode = async_mode

    def _base_measure(self, test_case: LLMTestCase):
        self.success = Scorer.exact_match_score(
            test_case.actual_output,
            test_case.expected_output
        )
        if self.success:
            self.score = 1
        else:
            self.score = 0
        return self.score

    def measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    async def a_measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    def is_successful(self):
        return bool(self.success)

    @property
    def name(self):
        return "ExactMatch"

def create_evaluation_dataset(df, llm, prompt_template):
    test_cases = []

    for _, row in df.iterrows():

        # Formatting prompt with the query from df
        prompt = prompt_template.format(query=row['Query'])

        # translating text to sql using llm
        result = llm.generate([[HumanMessage(content=prompt)]])

        actual_output = result.generations[0][0].text

        # Creating test cases and appending
        test_case = LLMTestCase(
            input=row['Query'],
            actual_output=actual_output,
            expected_output=row['Ground Truth'],
            context=[row['Query']]
        )
        test_cases.append(test_case)

    return EvaluationDataset(test_cases=test_cases)

def run_evaluation(df, llm, prompt_template):

    print("Creating evaluation dataset")
    dataset = create_evaluation_dataset(df, llm, prompt_template)

    print("\n Running evaluation metrics")
    # Checks if the generated SQL exactly matches the ground truth, useful for basic validation
    exact_match_metric = ExactMatchMetric()

    # Checks if the model generates non-existent tables/columns, using 0.3 which means say allowing some minor deviations in results
    hallucination_metric = HallucinationMetric(threshold=0.3)

    # Ensures the generated SQL actually addresses the input question, using 0.5 i.e results should be at least moderately relevant
    answer_relevancy_metric = AnswerRelevancyMetric(threshold=0.5)

    evaluation_results = evaluate(dataset, [
        exact_match_metric,
        hallucination_metric,
        answer_relevancy_metric
        ])

    pass_rates = aggregate_metric_pass_rates(evaluation_results.test_results)

    print("\nEvaluation Results:")
    print("=====================")
    for metric, pass_rate in pass_rates.items():
        print(f"{metric}: {pass_rate * 100:.2f}% pass rate")

    return evaluation_results, pass_rates

In [30]:
ZERO_SHOT_PROMPT_TEMPLATE = """Generate a valid SQL query for the following natural language instruction:
{query}
Only generate SQL code and nothing else."""

# Run evaluation
results, pass_rates = run_evaluation(df, llm, ZERO_SHOT_PROMPT_TEMPLATE)

Creating evaluation dataset

 Running evaluation metrics


✨ You're running DeepEval's latest Base Metric Metric! (using None, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Hallucination Metric! (using gpt-4o, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Answer Relevancy Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:14,  1.00test case/s]



Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation model: None, reason: None, error: None)
  - ✅ Hallucination (score: 0.0, threshold: 0.3, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the actual output is fully aligned with the context, with no contradictions present., error: None)
  - ✅ Answer Relevancy (score: 1.0, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The score is 1.00 because the response was perfectly relevant, directly addressing the question with no irrelevant information. Great job!, error: None)

For test case:

  - input: Show the total number of employees.
  - actual output: ```sql
SELECT COUNT(*) AS total_employees FROM employees;
```
  - expected output: SELECT COUNT(*) AS total_employees FROM employees;
  - context: ['Show the total number of employees.']
  - retrieval context: None


Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation m

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

Base Metric: 0.00% pass rate
Hallucination: 93.33% pass rate
Answer Relevancy: 100.00% pass rate



Evaluation Results:
Base Metric: 0.00% pass rate
Hallucination: 93.33% pass rate
Answer Relevancy: 100.00% pass rate


Good job evaluating the LLM and Prompt on the entire dataset that we have created! If you notice it isn't that good at this task still. What do you think we can do to improve the evaluation performance? What issues do you see with the way we perform evaluations at this point of time?


Following are some observations with current evaluation:

- **Exact Match metric** has 0% pass rate because it's too strict
- **Hallucination metric** has 93.33% pass rate  but failed on `Show the total number of orders placed each day last week` which is a False positive since the only difference is use of **CURDATE** and **CURRENT_DATE** between actual and expected query
- **Answer relevancy metric** has 100% pass rate which could be considered as less strict metric as it will only focus on the relevancy part of query but not on the execution part of it

# 3. Validating LLM Outputs using Guardrails

When building our LLM Application one difficult thing is the fact that LLM's are non-deterministic in their outputs. We are able to control different parameters within our LLM to mitigate this to a certain extent such as:
1. temperature / top_p
2. seed

You can read more about these parameters from OpenAI's documentation [here](https://platform.openai.com/docs/api-reference/chat/create).

But ensuring consistent outputs is still a problem. *This is where having guardrails comes into play.*

> Guardrails = safety controls for LLMs

Just like guardrails help prevent cars from going off the road, they can also help out with making sure LLM outputs are in line with set expectations.

![Road guardrails](https://drive.google.com/uc?id=1KZCfJs9Sf6ExqpphnTa4G-AaisxFBRWG)

Guardrails can take many different forms which differ in complexity:
1. **Security Guardrails**, ensuring there isn't any PII data in the response
2. **Compliance Guardrails**, ensuring that competitors are not talked about
3. **Safety Guardrails**, ensuring that there are not toxic responses
4. **Structural Guardrails**, ensuring that the response matches a specific JSON schema


A simple Guardrail for this situation could be to ensure that there are **no default error results incorporated within the product description**.

In our application we will also be communicating via REST APIs which leverage JSON to send and receive data between different applications.

**Frontend JSON sent to the Backend:**
```json
{
    "prompt": "Write a SQL query to find the names and email addresses of all users in the 'users' table who joined after January 1, 2020.",
    "temperature": 0.1,
    "max_tokens": 100
}
```
**Backend Response JSON sent back to the Frontend:**
```json
{
  "id": "cmpl-XYZ456",
  "object": "text_completion",
  "created": 1616517999,
  "model": "gpt-3.5-turbo",
  "prediction": {
    "text": "SELECT name, email FROM users WHERE join_date > '2020-01-01';"
  }
}
```
The response from OpenAI will be used to populate the `prediction` field within the main Backend Response where we will have an object that contains a `text` field which will then be used to contain the actual output from OpenAI.

In this scenario we would initially look for a **Structural Guardrail** that ensures that the text response from the LLM:
1. *is a string*
2. *is not empty*

A great library for this would be [Guardrails.ai](https://www.guardrailsai.com/) which allows us to leverage [Pydantic](https://docs.pydantic.dev/latest/) in order to easily define validators for our LLM responses.


In [31]:
!guardrails configure --token $GUARDRAILS_TOKEN

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
Enable anonymous metrics reporting? [Y/n]: Y
Do you wish to use remote inferencing? [Y/n]: Y

            Login successful.

            Get started by installing our RegexMatch validator:
            https://hub.guardrailsai.com/validator/guardrails_ai/regex_match

            You can install it by running:
            guardrails hub install hub://guardrails/regex_match

            Find more validators at https://hub.guardrailsai.com
            


In [32]:
from pydantic import BaseModel, Field

class LLMResponse(BaseModel):
  generated_sql: str = Field(description="Generated SQL from LLM")

In [34]:
import openai
from rich import print
from guardrails import Guard

# Add in your prompt here from before along with the query
# NOTE: gr.complete_json_suffix_v2 comes from guardrail where it will inject
#       additional tokens related to enforcing the Pydantic Model, to the main prompt.
ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE = """
Generate a valid SQL query for the following natural language instructions:

Instructions:
- Only generate a SELECT query
- End the query with a semicolon
- Do not include any explanations, only the SQL query

${query}

${gr.complete_json_suffix_v3}
"""

guard = Guard.from_pydantic(output_class=LLMResponse, prompt=ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE)

raw_llm_output, validated_output, *rest = guard(
    llm_api=openai.chat.completions.create,
    model=MODEL_NAME,
    prompt_params={
        "query": SAMPLE_QUERY
    },
)


In [35]:
# Shows what was sent to the LLM, including the guardrails formatting
print(guard.history.last.prompt)

Generate a valid SQL query for the following natural language instructions:

Instructions:
- Only generate a SELECT query
- End the query with a semicolon
- Do not include any explanations, only the SQL query

${query}

${gr.complete_json_suffix_v3}

In [36]:
# shows the raw LLM response
print(raw_llm_output)

{"generated_sql":"SELECT name FROM employees ORDER BY salary DESC LIMIT 1;"}

In [37]:
# Shows llm response object with generated_sql, it was successful
print(validated_output)

{'generated_sql': 'SELECT name FROM employees ORDER BY salary DESC LIMIT 1;'}

In [38]:
# We can view the history of our prompts in a nice tree like format
print(guard.history.last.tree)

Logs
└── ╭────────────────────────────────────────────────── Step 0 ───────────────────────────────────────────────────╮
    │ ╭──────────────────────────────────────────────── Prompt ─────────────────────────────────────────────────╮ │
    │ │                                                                                                         │ │
    │ │ Generate a valid SQL query for the following natural language instructions:                             │ │
    │ │                                                                                                         │ │
    │ │ Instructions:                                                                                           │ │
    │ │ - Only generate a SELECT query                                                                          │ │
    │ │ - End the query with a semicolon                                                                        │ │
    │ │ - Do not include any explanations, only the SQL query                                                   │ │
    │ │                                                                                                         │ │
    │ │ Select the name of the employee who has the highest salary                                              │ │
    │ │                                                                                                         │ │
    │ │                                                                                                         │ │
    │ │ Given below is JSON Schema that describes the information to extract from this document and the tags to │ │
    │ │ extract it into.                                                                                        │ │
    │ │                                                                                                         │ │
    │ │ {"properties": {"generated_sql": {"description": "Generated SQL from LLM", "title": "Generated Sql",    │ │
    │ │ "type": "string"}}, "required": ["generated_sql"], "type": "object", "title": "LLMResponse"}            │ │
    │ │                                                                                                         │ │
    │ │ ONLY return a valid JSON object (no other text is necessary). The JSON MUST conform to the JSON Schema, │ │
    │ │ including any types and format requests e.g. requests for lists, objects and specific types. Be correct │ │
    │ │ and concise. If you are unsure anywhere, try your best guess.                                           │ │
    │ │                                                                                                         │ │
    │ │ Here are examples of simple (JSON Schema, JSON) pairs that show the expected behavior:                  │ │
    │ │ - `{"type":"object","properties":{"foo":{"type":"string","format":"two-words lower-case"}}}` =>         │ │
    │ │ `{'foo': 'example one'}`                                                                                │ │
    │ │ -                                                                                                       │ │
    │ │ `{"type":"object","properties":{"bar":{"type":"array","items":{"type":"string","format":"upper-case"}}} │ │
    │ │ }` => `{"bar": ['STRING ONE', 'STRING TWO']}`                                                           │ │
    │ │ -                                                                                                       │ │
    │ │ `{"type":"object","properties":{"baz":{"type":"object","properties":{"foo":{"type":"string","format":"c │ │
    │ │ apitalize two-words"},"index":{"type":"integer","format":"1-indexed"}}}}}` => `{'baz': {'foo': 'Some    │ │
    │ │ String', 'index': 1}}`                                                                                  │ │
    │ │ -                                                                                                       │ │
    │ │ `{"type":"object","properties":{"bar":{"typ

## TODO: Adding More Guardrails
We have created a very basic Guardrail that leverages Pydantic Model's to ensure that our response fits a specifc format. Now let us try more advanced guardrails:
- Try creating **Bug Free SQL Code** leveraging the following [example](https://www.guardrailsai.com/docs/examples/syntax_error_free_sql).
- Feel free to add in more [guardrails via the Guardrails.ai Hub](https://hub.guardrailsai.com/) where you see fit for our use case. For example, we could ensure that there isn't any toxic language being returned from out LLM via using the [ToxicLanguage Validator](https://www.guardrailsai.com/docs/examples/toxic_language).
- How should we handle validation failures? One way would be to leverage reasking, where we reask the LLM to fix the output from a validation check failure. Within Guardrails.ai there are corrective measures we can take to handle validation errors as outlined [here](https://www.guardrailsai.com/docs/how_to_guides/rail#%EF%B8%8F-specifying-corrective-actions).



In [44]:
! guardrails hub install hub://guardrails/valid_sql --quiet

Installing hub://guardrails/valid_sql...
✅Successfully installed guardrails/valid_sql!




In [47]:
! pip install sqlvalidator -q

## Bug Free SQL Code

In [51]:
import openai
from rich import print
from guardrails import Guard
from guardrails.hub.guardrails.valid_sql.validator import ValidSQL

ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE = """
Generate a valid SQL query for the following natural language instructions:

Instructions:
- Only generate a SELECT query
- End the query with a semicolon
- Do not include any explanations, only the SQL query

${query}

${gr.complete_json_suffix_v3}
"""

# Defining output class with the ValidSQL validator
class LLMResponse(BaseModel):
    generated_sql: str = Field(description="Generate SQL for the given natural language instruction.", validators=[ValidSQL(on_fail="reask")])


guard = Guard.from_pydantic(output_class=LLMResponse, prompt=ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE)


raw_llm_output, validated_output, *rest = guard(
    llm_api=openai.chat.completions.create,
    model=MODEL_NAME,
    prompt_params={
        "query": SAMPLE_QUERY
    },
)

print(validated_output)

{'generated_sql': 'SELECT name FROM employees ORDER BY salary DESC LIMIT 1;'}

In [52]:
! guardrails hub install hub://guardrails/toxic_language --quiet

Installing hub://guardrails/toxic_language...
✅Successfully installed guardrails/toxic_language!





 ### Toxic Language Validator

In [54]:

import openai
from rich import print
from guardrails import Guard
from guardrails.hub.guardrails.toxic_language.validator import ToxicLanguage


ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE = """
Generate a valid SQL query for the following natural language instructions:

Instructions:
- Only generate a SELECT query
- End the query with a semicolon
- Do not include any explanations, only the SQL query

${query}

${gr.complete_json_suffix_v3}
"""

# Defining output class with the ValidSQL validator
class LLMResponse(BaseModel):
    generated_sql: str = Field(description="Generate SQL for the given natural language instruction.", validators=[ToxicLanguage(on_fail="reask")])


guard = Guard.from_pydantic(output_class=LLMResponse, prompt=ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE)


raw_llm_output, validated_output, *rest = guard(
    llm_api=openai.chat.completions.create,
    model=MODEL_NAME,
    prompt_params={
        "query": SAMPLE_QUERY
    },
)

print(validated_output)

{'generated_sql': 'SELECT name FROM employees ORDER BY salary DESC LIMIT 1;'}

### How should we handle validation failures?

In [55]:
# To automatically correct outputs containing toxic language, we will set on_fail='fix'
import openai
from rich import print
from guardrails import Guard
from guardrails.hub.guardrails.toxic_language.validator import ToxicLanguage


ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE = """
Generate a valid SQL query for the following natural language instructions:

Instructions:
- Only generate a SELECT query
- End the query with a semicolon
- Do not include any explanations, only the SQL query

${query}

${gr.complete_json_suffix_v3}
"""

# Defining output class with the ValidSQL validator
class LLMResponse(BaseModel):
    generated_sql: str = Field(description="Generate SQL for the given natural language instruction.", validators=[ToxicLanguage(on_fail="fix")])


guard = Guard.from_pydantic(output_class=LLMResponse, prompt=ZERO_SHOT_PROMPT_GUARDRAILS_TEMPLATE)


raw_llm_output, validated_output, *rest = guard(
    llm_api=openai.chat.completions.create,
    model=MODEL_NAME,
    prompt_params={
        "query": SAMPLE_QUERY
    },
)

print(validated_output)


{'generated_sql': 'SELECT name FROM employees ORDER BY salary DESC LIMIT 1;'}

Great job adding in robust guardrails, next we can now focus on making sure our system performs at peak performance. Let us try improving the results of the model using Prompt Engineering techniques making sure that we compare results from the baseline with the new improvements. Feel free to go wild and experiment with all kinds of techniques for this section!

## TODO: Evaluation Driven Development(EDD)

EDD is similar to [Test Driven Development(TDD)](https://martinfowler.com/bliki/TestDrivenDevelopment.html) except we apply it to building LLM-powered applications, where our test cases in this case are derived from the Golden dataset that we created earlier and the goal here is to iteratively ensure that our LLM application can achieve a **100% pass rate** across all the examples.

The main idea here is that if we currently have a 60% pass rate on the exact match metric, we should take an iterative approach towards increasing the pass rate. For example, the LLM could be failing on the following query `Display the department that has the highest average employee salary.`. In this section we will try to tune the prompt or other parts to ensure that the model is able to correctly generate the ground truth for it.

NOTE: As we mentioned before some items from the dataset could result in false negatives since multiple implementations are possible for the same query.

Here are a couple of things we can try out:
- Try out Few-Shot Prompting and [other techniques](https://www.promptingguide.ai/techniques) and see how it impacts our evaluation performance.
- Try tuning the parameters used for a model.
- Try out different LLM models from OpenAI, Anthropic, etc. if you have API keys for them.

You will have to setup an EvaluationDataset per each comparison since each run would have a different `actual_output` used within the LLMTestCase.
```python
from deepeval.evaluate import aggregate_metric_pass_rates
from deepeval.dataset import EvaluationDataset

dataset_run_1 = EvaluationDataset(test_cases=[...])
dataset_run_2 = EvaluationDataset(test_cases=[...])

evaluation_results_run_1 = evaluate(dataset_run_1, [exact_match_metric])
pass_rate_run_1 = aggregate_metric_pass_rates(evaluation_results_run_1.test_results)

evaluation_results_run_2 = evaluate(dataset_run_2, [exact_match_metric])
pass_rate_run_2 = aggregate_metric_pass_rates(evaluation_results_run_2.test_results)

assert pass_rate_run_2.get("ExactMatch") > pass_rate_run_1.get("ExactMatch"), \
  "Run 2 does not perform better than Run 1"
```


In [57]:
from deepeval.metrics.base_metric import BaseMetric
from deepeval.scorer import Scorer
from deepeval.test_case import LLMTestCase
from deepeval.dataset import EvaluationDataset
from langchain.schema import HumanMessage
from deepeval import evaluate
from deepeval.metrics import HallucinationMetric, AnswerRelevancyMetric
from deepeval.evaluate import aggregate_metric_pass_rates


class ExactMatchMetric(BaseMetric):
    def __init__(self, threshold=0.0, async_mode=True):
        self.threshold = threshold
        self.async_mode = async_mode

    def _base_measure(self, test_case: LLMTestCase):
        self.success = Scorer.exact_match_score(
            test_case.actual_output,
            test_case.expected_output
        )
        self.score = 1 if self.success else 0
        return self.score

    def measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    async def a_measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    def is_successful(self):
        return bool(self.success)

    @property
    def name(self):
        return "ExactMatch"


def create_evaluation_dataset(df, llm, prompt_template):
    test_cases = []
    for _, row in df.iterrows():
        prompt = prompt_template.format(query=row['Query'])
        result = llm.generate([[HumanMessage(content=prompt)]])
        actual_output = result.generations[0][0].text
        test_case = LLMTestCase(
            input=row['Query'],
            actual_output=actual_output,
            expected_output=row['Ground Truth'],
            context=[row['Query']]
        )
        test_cases.append(test_case)
    return EvaluationDataset(test_cases=test_cases)



ZERO_SHOT_PROMPT_TEMPLATE = """Generate a valid SQL query for the following natural language instruction:
{query}
Only generate SQL code and nothing else."""


FEW_SHOT_PROMPT_TEMPLATE = """
Here are some examples of natural language instructions and their SQL equivalents:
1. Instruction: "Show the name and salary of employees who earn more than $100,000."
   SQL: "SELECT name, salary FROM employees WHERE salary > 100000;"
2. Instruction: "List the departments along with the number of employees in each."
   SQL: "SELECT department, COUNT(*) FROM employees GROUP BY department;"

Now generate a valid SQL query for the following instruction:
{query}
Only provide the SQL code.
"""


def run_evaluation(df, llm, prompt_template, prompt_name):
    print(f"Creating evaluation dataset for {prompt_name} prompting...")
    dataset = create_evaluation_dataset(df, llm, prompt_template)

    print(f"\nRunning evaluation metrics for {prompt_name} prompting...")
    exact_match_metric = ExactMatchMetric()
    hallucination_metric = HallucinationMetric(threshold=0.3)
    answer_relevancy_metric = AnswerRelevancyMetric(threshold=0.5)

    evaluation_results = evaluate(dataset, [
        exact_match_metric,
        hallucination_metric,
        answer_relevancy_metric
    ])

    pass_rates = aggregate_metric_pass_rates(evaluation_results.test_results)

    print(f"\nEvaluation Results for {prompt_name} Prompting:")
    print("===================================================")
    for metric, pass_rate in pass_rates.items():
        print(f"{metric}: {pass_rate * 100:.2f}% pass rate")

    return evaluation_results, pass_rates

In [58]:
results_zero_shot, pass_rates_zero_shot = run_evaluation(df, llm, ZERO_SHOT_PROMPT_TEMPLATE, "Zero-Shot")

Creating evaluation dataset for Zero-Shot prompting...

Running evaluation metrics for Zero-Shot prompting...

✨ You're running DeepEval's latest Base Metric Metric! (using None, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Hallucination Metric! (using gpt-4o, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Answer Relevancy Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:17,  1.16s/test case]




Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation model: None, reason: None, error: None)
  - ✅ Hallucination (score: 0.0, threshold: 0.3, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the actual output fully aligns with the context, with no contradictions present., error: None)
  - ✅ Answer Relevancy (score: 1.0, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The score is 1.00 because the response is perfectly relevant and focused, providing exactly the information requested without any irrelevant details. Great job!, error: None)

For test case:

  - input: List the names and emails of staff in the IT department.
  - actual output: ```sql
SELECT name, email FROM staff WHERE department = 'IT';
```
  - expected output: SELECT name, email FROM employees WHERE department = 'IT';
  - context: ['List the names and emails of staff in the IT department.']
  - retrieval context: None


Metrics Summary

  

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

Base Metric: 0.00% pass rate
Hallucination: 93.33% pass rate
Answer Relevancy: 93.33% pass rate




Evaluation Results for Zero-Shot Prompting:

===================================================

Base Metric: 0.00% pass rate

Hallucination: 93.33% pass rate

Answer Relevancy: 93.33% pass rate

In [59]:
results_few_shot, pass_rates_few_shot = run_evaluation(df, llm, FEW_SHOT_PROMPT_TEMPLATE, "Few-Shot")

Creating evaluation dataset for Few-Shot prompting...

Running evaluation metrics for Few-Shot prompting...

✨ You're running DeepEval's latest Base Metric Metric! (using None, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Hallucination Metric! (using gpt-4o, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Answer Relevancy Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:16,  1.10s/test case]



Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation model: None, reason: None, error: None)
  - ✅ Hallucination (score: 0.0, threshold: 0.3, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the actual output fully aligns with the provided context, with no contradictions present., error: None)
  - ✅ Answer Relevancy (score: 1.0, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The score is 1.00 because the output is perfectly relevant, addressing the input question directly without any irrelevant statements. Great job!, error: None)

For test case:

  - input: What are the titles of the top 5 selling books?
  - actual output: ```sql
SELECT title FROM books ORDER BY sales DESC LIMIT 5;
```
  - expected output: SELECT title FROM books ORDER BY sales DESC LIMIT 5;
  - context: ['What are the titles of the top 5 selling books?']
  - retrieval context: None


Metrics Summary

  - ❌ Base Metric (score: 0.0, thre

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

Base Metric: 0.00% pass rate
Hallucination: 100.00% pass rate
Answer Relevancy: 100.00% pass rate




Evaluation Results for Few-Shot Prompting:

===================================================

Base Metric: 0.00% pass rate

Hallucination: 100.00% pass rate

Answer Relevancy: 100.00% pass rate

In [60]:
print("\nComparison of Prompting Techniques:")
for metric in pass_rates_zero_shot.keys():
    zero_shot_rate = pass_rates_zero_shot[metric] * 100
    few_shot_rate = pass_rates_few_shot[metric] * 100
    print(f"{metric} Pass Rate:")
    print(f"  Zero-Shot: {zero_shot_rate:.2f}%")
    print(f"  Few-Shot: {few_shot_rate:.2f}%")

Comparison of Prompting Techniques:

Base Metric Pass Rate:

Zero-Shot: 0.00%

Few-Shot: 0.00%

Hallucination Pass Rate:

Zero-Shot: 93.33%

Few-Shot: 100.00%

Answer Relevancy Pass Rate:

Zero-Shot: 93.33%

Few-Shot: 100.00%

- Higher pass rates for hallucination with few shot prompting as compared to zero shot prompting

### Tuning the parameters used for a model.

In [64]:
from deepeval.metrics.base_metric import BaseMetric
from deepeval.scorer import Scorer
from deepeval.test_case import LLMTestCase
from deepeval.dataset import EvaluationDataset
from langchain.schema import HumanMessage
from deepeval import evaluate
from deepeval.metrics import HallucinationMetric, AnswerRelevancyMetric
from deepeval.evaluate import aggregate_metric_pass_rates


class ExactMatchMetric(BaseMetric):
    def __init__(self, threshold=0.0, async_mode=True):
        self.threshold = threshold
        self.async_mode = async_mode

    def _base_measure(self, test_case: LLMTestCase):
        self.success = Scorer.exact_match_score(
            test_case.actual_output,
            test_case.expected_output
        )
        self.score = 1 if self.success else 0
        return self.score

    def measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    async def a_measure(self, test_case: LLMTestCase):
        self._base_measure(test_case=test_case)

    def is_successful(self):
        return bool(self.success)

    @property
    def name(self):
        return "ExactMatch"


def create_evaluation_dataset(df, llm, prompt_template, llm_config):
    test_cases = []
    for _, row in df.iterrows():
        prompt = prompt_template.format(query=row['Query'])
        result = llm.generate([[HumanMessage(content=prompt)]])
        actual_output = result.generations[0][0].text
        test_case = LLMTestCase(
            input=row['Query'],
            actual_output=actual_output,
            expected_output=row['Ground Truth'],
            context=[row['Query']]
        )
        test_cases.append(test_case)
    return EvaluationDataset(test_cases=test_cases)



ZERO_SHOT_PROMPT_TEMPLATE = """Generate a valid SQL query for the following natural language instruction:
{query}
Only generate SQL code and nothing else."""


def run_evaluation(df, llm, prompt_template, llm_config):
    print("Creating evaluation dataset...")
    dataset = create_evaluation_dataset(df, llm, prompt_template, llm_config)

    print("\nRunning evaluation metrics...")
    exact_match_metric = ExactMatchMetric()
    hallucination_metric = HallucinationMetric(threshold=0.3)
    answer_relevancy_metric = AnswerRelevancyMetric(threshold=0.5)

    evaluation_results = evaluate(dataset, [
        exact_match_metric,
        hallucination_metric,
        answer_relevancy_metric
    ])

    pass_rates = aggregate_metric_pass_rates(evaluation_results.test_results)

    print("\nEvaluation Results:")
    print("======================")
    for metric, pass_rate in pass_rates.items():
        print(f"{metric}: {pass_rate * 100:.2f}% pass rate")

    return evaluation_results, pass_rates


# Configuration 1: Lower temperature for more deterministic output
llm_config_1 = {
    "temperature": 0.0,
    "max_tokens": 150,
}

# Configuration 2: Higher temperature for more creative output
llm_config_2 = {
    "temperature": 0.8,
    "max_tokens": 150
}

In [65]:
results_config_1, pass_rates_config_1 = run_evaluation(df, llm, ZERO_SHOT_PROMPT_TEMPLATE, llm_config_1)

Creating evaluation dataset...

Running evaluation metrics...

✨ You're running DeepEval's latest Base Metric Metric! (using None, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Hallucination Metric! (using gpt-4o, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Answer Relevancy Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:19,  1.30s/test case]



Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation model: None, reason: None, error: None)
  - ✅ Hallucination (score: 0.0, threshold: 0.3, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the actual output aligns perfectly with the context without any contradictions, indicating no hallucination., error: None)
  - ❌ Answer Relevancy (score: 0.0, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the output provided a SQL query, which does not directly answer the question about the titles of the top 5 selling books., error: None)

For test case:

  - input: What are the titles of the top 5 selling books?
  - actual output: ```sql
SELECT title 
FROM books 
ORDER BY sales DESC 
LIMIT 5;
```
  - expected output: SELECT title FROM books ORDER BY sales DESC LIMIT 5;
  - context: ['What are the titles of the top 5 selling books?']
  - retrieval context: None


Metrics Summary

  - ❌ Base

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

Base Metric: 0.00% pass rate
Hallucination: 93.33% pass rate
Answer Relevancy: 93.33% pass rate




Evaluation Results:

======================

Base Metric: 0.00% pass rate

Hallucination: 93.33% pass rate

Answer Relevancy: 93.33% pass rate

In [66]:
results_config_2, pass_rates_config_2 = run_evaluation(df, llm, ZERO_SHOT_PROMPT_TEMPLATE, llm_config_2)

Creating evaluation dataset...

Running evaluation metrics...

✨ You're running DeepEval's latest Base Metric Metric! (using None, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Hallucination Metric! (using gpt-4o, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Answer Relevancy Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:16,  1.11s/test case]



Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluation model: None, reason: None, error: None)
  - ✅ Hallucination (score: 0.0, threshold: 0.3, strict: False, evaluation model: gpt-4o, reason: The score is 0.00 because the actual output fully aligns with the context without any contradictions., error: None)
  - ✅ Answer Relevancy (score: 1.0, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The score is 1.00 because the response is perfectly relevant and concise, directly addressing the request without any irrelevant information. Great job!, error: None)

For test case:

  - input: Show the total number of employees.
  - actual output: ```sql
SELECT COUNT(*) AS total_employees FROM employees;
```
  - expected output: SELECT COUNT(*) AS total_employees FROM employees;
  - context: ['Show the total number of employees.']
  - retrieval context: None


Metrics Summary

  - ❌ Base Metric (score: 0.0, threshold: 0.0, strict: False, evaluat

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

Base Metric: 0.00% pass rate
Hallucination: 93.33% pass rate
Answer Relevancy: 100.00% pass rate




Evaluation Results:

======================

Base Metric: 0.00% pass rate

Hallucination: 93.33% pass rate

Answer Relevancy: 100.00% pass rate

In [67]:
print("\nComparison of Parameter Configurations:")
print("===========================================")
for metric in pass_rates_config_1.keys():
    config_1_rate = pass_rates_config_1[metric] * 100
    config_2_rate = pass_rates_config_2[metric] * 100
    print(f"{metric} Pass Rate:")
    print(f"  Config 1: {config_1_rate:.2f}%")
    print(f"  Config 2: {config_2_rate:.2f}%")

Comparison of Parameter Configurations:

===========================================

Base Metric Pass Rate:

Config 1: 0.00%

Config 2: 0.00%

Hallucination Pass Rate:

Config 1: 93.33%

Config 2: 93.33%

Answer Relevancy Pass Rate:

Config 1: 93.33%

Config 2: 100.00%

LLM conf 2 achieves a pass rate of 100%, while llm conf 1 has a slightly lower pass rate of 93.33% for **answer relevancy**

## TODO: Evaluating SQL Code Using a Stronger LLM

Usually we would prioritise latency during inference using LLMs for most use cases, LLMs such as `gpt3.5-turbo` are much faster than `gpt4` variants while performing worse in general across benchmarks. What if we could also use a bigger and stronger LLM to aid with evaluations as well? There has been a rise of techniques doing just that where a smaller and faster LLM is evaluated by a bigger and stronger LLM. This process helps automate the evaluation part of any LLM-powered application! This is kinda of like having a dream within a dream just like in the movie *Inception*.

![Dream within a dream from Inception](https://drive.google.com/uc?id=1EC7rNg_LCLjIJ9SRyQpGVgtplOHZllDm)


**NOTE:** LLMs are very powerful and can help automate so many parts of the process but becareful not to be too reliant on them, at the end of the day they aren't perfect and hallucinate a lot. Therefore, its vital that you always keep a human-in-the-loop to verify things as a final seal of approval.


### Introducing G-Eval

[G-Eval](https://arxiv.org/abs/2303.16634) is a method that utilizes LLMs alongside a Chain-of-Thought (CoT) and form-filling approach to assess the outputs of LLMs. Initially, a task outline and evaluation parameters are provided to an LLM, which is then requested to create a CoT detailing the evaluation steps. For assessing the coherence of news summaries, the prompt, CoT, news article, and summary are merged, and the LLM is instructed to produce a score ranging from 1 to 5. Subsequently, the output token probabilities generated by the LLM are used to standardize this score, with a weighted summation taken as the ultimate outcome.

It was observed that using GPT-4 as an evaluator resulted in a high Spearman correlation with human assessments (0.514), surpassing all former methods. In the realm of summarization, it exceeded all prior state-of-the-art (SOTA) evaluators in the SummEval benchmark, which evaluates coherence, consistency, fluency, and relevance.


![G-Eval](https://drive.google.com/uc?id=1dq1BAjLY_Es5r-UiD96ZnFlxWTYcNv8g)

deepeval already comes out of the box with support for using G-Eval [here](https://docs.confident-ai.com/docs/metrics-llm-evals) where you can very easily just create a new metric and run it against your EvaluationDataset:

```python
from deepeval.metrics import GEval
from deepeval.test_case import LLMTestCaseParams

correctness_metric = GEval(
    name="Correctness",
    criteria="Determine whether the actual output is factually correct based on the expected output.",
    # NOTE: you can only provide either criteria or evaluation_steps, and not both
    evaluation_steps=[
        "Check whether the facts in 'actual output' contradicts any facts in 'expected output'",
        "You should also heavily penalize omission of detail",
        "Vague language, or contradicting OPINIONS, are OK"
    ],
    evaluation_params=[LLMTestCaseParams.INPUT, LLMTestCaseParams.ACTUAL_OUTPUT],
)
```
Try creating a G-Eval metric for SQL Code Generation and running it through our EvaluationDataset like before.


**NOTE:** The API Key provided to you only allows access to `gpt-4o-mini`, we can still use it either way but in the real world you would ideally use a more powerful model for this task. If you were to use `gpt-4o` as your base model then you probably don't have another *stronger* LLM to use therefore this also mimics that kind of constraint.

In [68]:
from deepeval import evaluate
from deepeval.test_case import LLMTestCase
from deepeval.dataset import EvaluationDataset

def create_evaluation_dataset(df, llm, prompt_template):
    test_cases = []

    for _, row in df.iterrows():

        # Formatting prompt with the query from df
        prompt = prompt_template.format(query=row['Query'])

        # translating text to sql using llm
        result = llm.generate([[HumanMessage(content=prompt)]])

        actual_output = result.generations[0][0].text

        # Creating test cases and appending
        test_case = LLMTestCase(
            input=row['Query'],
            actual_output=actual_output,
            expected_output=row['Ground Truth'],
            context=[row['Query']]
        )
        test_cases.append(test_case)

    return EvaluationDataset(test_cases=test_cases)


ZERO_SHOT_PROMPT_TEMPLATE = """Generate a valid SQL query for the following natural language instruction:
{query}
Only generate SQL code and nothing else."""

In [69]:
evaluation_dataset = create_evaluation_dataset(df,llm,ZERO_SHOT_PROMPT_TEMPLATE)

In [70]:
from deepeval.metrics import GEval
from deepeval.test_case import LLMTestCaseParams

sql_correctness_metric = GEval(
    name="SQLCorrectness",
    criteria="Evaluate whether the generated SQL query correctly implements the intended functionality as described in the input.",
    evaluation_steps=[
        "Ensure that the query syntax is correct and adheres to SQL standards.",
        "Check for the use of appropriate SQL clauses and functions relevant to the task."
    ],
    evaluation_params=[LLMTestCaseParams.INPUT, LLMTestCaseParams.ACTUAL_OUTPUT],
)


In [71]:
evaluation_results = evaluate(evaluation_dataset, [sql_correctness_metric])
pass_rates = aggregate_metric_pass_rates(evaluation_results.test_results)

print("\nEvaluation Results:")
print("=====================")
for metric, pass_rate in pass_rates.items():
    print(f"{metric}: {pass_rate * 100:.2f}% pass rate")

✨ You're running DeepEval's latest SQLCorrectness (GEval) Metric! (using gpt-4o, strict=False, async_mode=True)...

Event loop is already running. Applying nest_asyncio patch to allow async execution...


Evaluating 15 test case(s) in parallel: |██████████|100% (15/15) [Time Taken: 00:07,  1.99test case/s]



Metrics Summary

  - ✅ SQLCorrectness (GEval) (score: 0.9999999999999998, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The SQL query correctly uses ORDER BY and LIMIT clauses to find the product with the least quantity in stock, adhering to SQL standards., error: None)

For test case:

  - input: Which product has the least quantity in stock?
  - actual output: ```sql
SELECT product_name 
FROM products 
ORDER BY quantity_in_stock ASC 
LIMIT 1;
```
  - expected output: SELECT product_name FROM products ORDER BY quantity_in_stock ASC LIMIT 1;
  - context: ['Which product has the least quantity in stock?']
  - retrieval context: None


Metrics Summary

  - ✅ SQLCorrectness (GEval) (score: 0.9979667646577666, threshold: 0.5, strict: False, evaluation model: gpt-4o, reason: The SQL query correctly uses the MAX function and the WHERE clause to find the highest salary in the Finance department, adhering to SQL standards., error: None)

For test case:

  - input: Find the

✓ Tests finished 🎉! Run 'deepeval login' to save and analyze evaluation results on Confident AI. 
‼️  Friendly reminder 😇: You can also run evaluations with ALL of deepeval's metrics directly on Confident AI 
instead.



Overall Metric Pass Rates

SQLCorrectness (GEval): 100.00% pass rate




Evaluation Results:

=====================

SQLCorrectness (GEval): 100.00% pass rate

Great job on getting G-Eval to work with our dataset! What do you think about using LLMs to evaluate another LLM vs just using traditional Machine Learning metrics? What are the Pros/Cons of it and where do you think it would work better or worse in?

**ANSWER**
- Traditional ML metrics works better for structured tasks in metrics like Precision, Recall and AUC for assessment while LLMs generate open ended responses across an infinite output space.

**Pros**
- Semantic Understanding, Context Awareness
- Scalabilty, Consistency and Flexibility

**Cons**
- Bias, Reliability, Resource Intensive and Verification Challenges

# 4. Extra Credit: Optional Tasks

- Try out implementing the Execution Score from the Snowflake team and comparing the results of it to Exact Match Metric and G-Eval, this [blog](https://medium.com/snowflake/inside-snowflake-building-the-most-powerful-sql-llm-in-the-world-1a33b3ee0d37) from the Snowflake Team covers how you can use LLMs to assists with evaluation.
- If you use SQL enough you would know that there are multiple ways in which we can answer a particular query especially if you had to update your ground truth earlier. With that you can try out other frameworks such as [sql-eval](https://github.com/defog-ai/sql-eval) which provides mo like we did before but more importanly also includes code execution into context when evaluating LLMs.
- Guardrails do not just apply to the output from an LLM, instead we can also apply them to the actual input that a user feeds into the LLM. You can try out **Detecting Toxicity**, **Restricting the input to just be related to SQL code generation**, **Limiting the amount of tokens used**, etc.
- Our BASIC_GROUND_TRUTH_DATASET is quite small, see how you can add in more examples using your own domain knowledge or you could even leverage an LLM to do so via synthetic data generation which is supported within deepeval [here](https://docs.confident-ai.com/docs/evaluation-datasets-synthetic-data). Remember not all users will put in queries that can be answered with SQL code. For example, if the user asks `Who will win the next US Presidential Election?` you definitely don't want to answer it at all!
